contributors/design-proposals/storage/volume-topology-scheduling.md

```diff
@@ -266,41 +304,368 @@ const (
+* "zone: us-central1-b" and "rack: rack2"
+
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  Name: something-fancy
+provisioner: rack-based-provisioner
+parameters:
+allowedTopologies:
+  requiredTopologies:
+  - matchNodeLabels:
+      zone: us-central1-a
+      rack: rack1
+  - matchNodeLabels:
+      zone: us-central1-b
+      rack: rack1
+  - matchNodeLabels:
+      zone: us-central1-b
+      rack: rack2
+```
+
+### Storage Provider Capacity Reporting
+Some volume types have capacity limitations per topology domain. For example,
+local volumes have limited capacity per node. Provisioners for these volume
+types can report these capacity limitations through a new StorageClass status
+field.
+
+```
+type StorageClass struct {
+    ...
+
+    Status *StorageClassStatus
+}
+
+type StorageClassStatus struct {
+    Capacity *StorageClassCapacity
+}
+
+type StorageClassCapacity struct {
+    // The node label key that the provisioner operates on
+    TopologyKey string
+    // Capacities specifies the total capacity available
+    // (free and used) for provisioning at each value of the
+    // TopologyKey
+    Capacities map[string]resource.Quantity
+}
+```
+
+If no StorageClassCapacity is specified, then there are no capacity
+limitations.
+
+Capacity reporting is limited to a single topology key. `TopologyKey`
+specifies the label key that represents that domain. Capacities is
+a map with key equal to the label value, and value is the capacity for that
+specific domain. Capacity is the total capacity available, including
+capacity that has already been provisioned. A missing map entry for a topology
+domain will be treated as zero capacity.
+
+A scheduler predicate will use the reported capacity to determine if dynamic
+provisioning is possible for a given node, taking into account already
+provisioned capacity.
+
+Downsides of this approach include:
+* Multiple topology keys is not supported for capacity reporting. We have not
+  seen such use cases so far.
+* There is no common mechanism to report capacity. Each provisioner has to
+  update their own fields in the StorageClass if necessary. CSI could potentially
+  address this.
+* There is a gap during cluster initialization where the StorageClass object
+  has been created, and the provisioner has not reported capacity yet. In this
+  scenario, the provisioner should fail provisioning until it has reported its
+  capacity and not process any PVCs that have not been updated with the chosen
+  topologyKey value.
+
+#### Local Volume Example
+This example specifies that node1 has 100Gi, and node2 has 200Gi available
+for dynamic provisioning. All other nodes have 0Gi available for dynamic
+provisioning.
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: standard
+...
+status:
+  capacity:
+    topologyKey: kubernetes.io/hostname
+    capacities:
+      node1: 100Gi
+      node2: 200Gi
+```
+
+TODO: should this be in the local volume spec instead?
```

Comment:
How this can be in local PV spec? We shouldn't put capacity of a node into all local PVs that are the node.